File: src/main/java/edu/kit/iti/formal/stvs/logic/io/xml/XmlConfigImporter.java
package edu.kit.iti.formal.stvs.logic.io.xml;

import edu.kit.iti.formal.stvs.logic.io.ImportException;
import edu.kit.iti.formal.stvs.model.config.GlobalConfig;

import java.io.File;
import java.net.URISyntaxException;
import javax.xml.bind.JAXBContext;
import javax.xml.bind.JAXBElement;
import javax.xml.bind.JAXBException;
import javax.xml.bind.Unmarshaller;

import org.w3c.dom.Node;

/**
 * This class provides the functionality to import a {@link GlobalConfig} from a xml node.
 *
 * @author Benjamin Alt
 */
public class  XmlConfigImporter extends XmlImporter<GlobalConfig> {

  /**
   * Imports a {@link GlobalConfig} from a xml {@link Node}.
   *
   * @param source Xml node that should be imported
   * @return Imported config
   * @throws ImportException Exception while importing
   */
  @Override
  public GlobalConfig doImportFromXmlNode(Node source) throws ImportException {
    try {
      JAXBContext jaxbContext = JAXBContext.newInstance(ObjectFactory.class);
      Unmarshaller jaxbUnmarshaller = jaxbContext.createUnmarshaller();
      Config importedConfig = ((JAXBElement<Config>) jaxbUnmarshaller.unmarshal(source)).getValue();
      GlobalConfig config = new GlobalConfig();
      setGeneralSettings(importedConfig.getGeneral(), config);
      setEditorSettings(importedConfig.getEditor(), config);
      setDependencies(importedConfig.getDependencies(), config);
      return config;
    } catch (JAXBException e) {
      throw new ImportException(e);
    }
  }

  /**
   * Set the dependencies section of a given {@link GlobalConfig} according to an imported
   * {@link edu.kit.iti.formal.stvs.logic.io.xml.Config.Dependencies} object.
   * @param deps The imported dependencies
   * @param config The config to modify
   */
  private void setDependencies(Config.Dependencies deps, GlobalConfig config) {
    if (deps != null) {
      if (deps.getGetetaCommand() != null) {
        config.setGetetaCommand(deps.getGetetaCommand());
      }
      if (deps.getNuxmvFilename() != null) {
        config.setNuxmvFilename(deps.getNuxmvFilename());
      }
      if (deps.getZ3Path() != null) {
        config.setZ3Path(deps.getZ3Path());
      }
    }
  }

  /**
   * Set the "general" section of a given {@link GlobalConfig} according to an imported
   * {@link edu.kit.iti.formal.stvs.logic.io.xml.Config.General} object.
   * @param general The imported {@link edu.kit.iti.formal.stvs.logic.io.xml.Config.General} object
   * @param config The config to modify
   */
  private void setGeneralSettings(Config.General general, GlobalConfig config) {
    if (general != null) {
      if (general.getUiLanguage() != null) {
        config.setUiLanguage(general.getUiLanguage());
      }
      if (general.getSimulationTimeout() != null) {
        config.setSimulationTimeout(general.getSimulationTimeout().intValue());
      }
      if (general.getVerificationTimeout() != null) {
        config.setVerificationTimeout(general.getVerificationTimeout().intValue());
      }
      if (general.getWindowSize().getHeight() != null) {
        config.setWindowHeight(general.getWindowSize().getHeight().intValue());
      }
      if (general.getWindowSize().getWidth() != null) {
        config.setWindowWidth(general.getWindowSize().getWidth().intValue());
      }
      if (general.getWindowSize().isMaximized() != null) {
        config.setWindowMaximized(general.getWindowSize().isMaximized());
      }
      if (general.getMaxLineRollout() != null) {
        config.setMaxLineRollout(general.getMaxLineRollout().intValue());
      }
    }
  }

  /**
   * Set the "editor" section of a given {@link GlobalConfig} according to an imported
   * {@link edu.kit.iti.formal.stvs.logic.io.xml.Config.Editor} object.
   * @param editor The imported {@link edu.kit.iti.formal.stvs.logic.io.xml.Config.Editor} object
   * @param config The config to modify
   */
  private void setEditorSettings(Config.Editor editor, GlobalConfig config) {
    if (editor != null) {
      if (editor.getFontSize() != null) {
        config.setEditorFontSize(editor.getFontSize().intValue());
      }
      if (editor.getFontFamily() != null) {
        config.setEditorFontFamily(editor.getFontFamily());
      }
      config.setShowLineNumbers(editor.isShowLineNumbers());
    }
  }

  @Override
  protected String getXsdFilePath() throws URISyntaxException {
    File xsdFile = new File(this.getClass().getResource("/fileFormats/config.xsd").toURI());
    return xsdFile.getAbsolutePath();
  }
}

File: stvs/src/main/kotlin/edu/kit/iti/formal/stvs/logic/io/xml/XmlConfigImporter.kt
/* *****************************************************************
 * This file belongs to verifaps-lib (https://verifaps.github.io).
 * SPDX-License-Header: GPL-3.0-or-later
 * 
 * This program isType free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program isType distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a clone of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * *****************************************************************/
package edu.kit.iti.formal.stvs.logic.io.xml

import edu.kit.iti.formal.stvs.logic.io.ImportException
import edu.kit.iti.formal.stvs.model.config.GlobalConfig
import org.jdom2.Element
import org.w3c.dom.Node
import java.net.URL

/**
 * This class provides the functionality to import a [GlobalConfig] from a xml node.
 *
 * @author Benjamin Alt
 */
class XmlConfigImporter : XmlImporter<GlobalConfig>() {
    /**
     * Imports a [GlobalConfig] from a xml [Node].
     *
     * @param source Xml node that should be imported
     * @return Imported config
     * @throws ImportException Exception while importing
     */
    @Throws(ImportException::class)
    override fun doImportFromXmlNode(source: Element): GlobalConfig {
        val config = GlobalConfig()
        setGeneralSettings(source.getChild("general"), config)
        setEditorSettings(source.getChild("editor"), config)
        setDependencies(source.getChild("dependencies"), config)
        return config
    }

    /**
     * Set the dependencies section of a given [GlobalConfig] according to an imported
     * [edu.kit.iti.formal.stvs.logic.io.xml.Config.Dependencies] object.
     *
     * @param deps   The imported dependencies
     * @param config The config to modify
     */
    private fun setDependencies(deps: Element?, config: GlobalConfig) {
        deps?.let {
            deps.getChildText("nuxmvFilename")?.let {
                config.nuxmvFilename = it
            }
            deps.getChildText("z3Path")?.let {
                config.z3Path = it
            }
        }
    }

    /**
     * Set the "general" section of a given [GlobalConfig] according to an imported
     * [edu.kit.iti.formal.stvs.logic.io.xml.Config.General] object.
     *
     * @param general The imported [edu.kit.iti.formal.stvs.logic.io.xml.Config.General] object
     * @param config  The config to modify
     */
    private fun setGeneralSettings(general: Element?, config: GlobalConfig) {
        general?.let {
            general.getChildText("simulationTimeout")?.let {
                config.simulationTimeout = it.toInt()
            }

            general.getChildText("verificationTimeout")?.let {
                config.verificationTimeout = it.toLong()
            }

            general.getChild("windowSize")?.let { windowSize ->
                windowSize.getChildText("height")?.let {
                    config.windowHeight = it.toInt()
                }
                windowSize.getChildText("width")?.let {
                    config.windowWidth = it.toInt()
                }
                windowSize.getChildText("maximized")?.let {
                    config.windowMaximized = it.toBoolean()
                }
            }
            general.getChildText("maxLineRollout")?.let {
                config.maxLineRollout = it.toInt()
            }
        }
    }

    /**
     * Set the "editor" section of a given [GlobalConfig] according to an imported
     * [edu.kit.iti.formal.stvs.logic.io.xml.Config.Editor] object.
     *
     * @param editor The imported [edu.kit.iti.formal.stvs.logic.io.xml.Config.Editor] object
     * @param config The config to modify
     */
    private fun setEditorSettings(editor: Element?, config: GlobalConfig) {
        editor?.let {
            editor.getChildText("fontSize")?.let {
                config.editorFontSize = it.toInt()
            }
            editor.getChildText("fontFamily")?.let {
                config.editorFontFamily = it
            }
        }
    }

    override val xsdResource: URL?
        get() = javaClass.getResource("/fileFormats/stvs-1.0.xsd")
}